### ISSUE TITLE
Hamcrest old version causes: java.lang.NoSuchMethodError: No static method allOf 

### ISSUE THREAD

**Issue by rsavin** [AdevintaSpain/Barista#357]
*Please consider making a Pull Request if you are capable of doing so.*

**Library Version:**

3.5.0
  
**Describe the Bug:**

When I use the `assertListItemCount(R.id.list, 5)` method and some other methods from Barista I get the following exception:

```
java.lang.NoSuchMethodError: No static method allOf(Lorg/hamcrest/Matcher;Lorg/hamcrest/Matcher;Lorg/hamcrest/Matcher;)Lorg/hamcrest/Matcher; in class Lorg/hamcrest/CoreMatchers; or its super classes (declaration of 'org.hamcrest.CoreMatchers' appears in /data/app/com.myapp.android.debug.test-teuB7mkareqg4fXqftwOghQ==/base.apk!classes3.dex)
at com.schibsted.spain.barista.interaction.BaristaListInteractions.findRecyclerMatcher(BaristaListInteractions.kt:118)
at com.schibsted.spain.barista.assertion.BaristaListAssertions.assertListItemCount(BaristaListAssertions.kt:33)
...
at java.lang.reflect.Method.invoke(Native Method)
at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
at androidx.test.internal.runner.junit4.statement.RunBefores.evaluate(RunBefores.java:80)
at androidx.test.internal.runner.junit4.statement.RunAfters.evaluate(RunAfters.java:61)
at androidx.test.rule.ActivityTestRule$ActivityStatement.evaluate(ActivityTestRule.java:531)
at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:288)
at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:282)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)
at java.lang.Thread.run(Thread.java:764)
```

I did some investigation and the problem is that we have Hamcrest v2.2 in our project and Barista uses Hamcrest v1.3 which comes transitively from Espresso dependencies. If I downgrade Hamcrest to v1.3 in the project dependencies it works fine.

Is it possible to start using Hamcrest v2.2 in Barista to make it possible to use in project with the updated library? I can open a PR if it's fine for you to get updated.

**Steps to reproduce the bug:**

**Expected Behavior:**

**Comment by Sloy** [comment]
@rsavin Hi Roman! Thanks for the report. 
Are you able to reproduce this issue in the sample project from this repository, or any other project? How are you adding the Hamcrest dependency?
I tried adding `androidTestImplementation("org.hamcrest:hamcrest:2.2")` to the sample and to another internal project, and the build just fails with errors like

```
Duplicate class org.hamcrest.BaseDescription found in modules hamcrest-2.2.jar (org.hamcrest:hamcrest:2.2) and hamcrest-core-1.3.jar (org.hamcrest:hamcrest-core:1.3)
```

If both versions are not compatible, we won't be able to update Hamcrest until Espresso does it first. And even if the build would work, we shouldn't because we might break other projects out there not using 2.2.

**Comment by Sloy** [comment]
Closing for inactivity

**Comment by shuhart** [comment]
We got the same issue. Espresso 3.4.0 uses hamcrest-library 2.2, which causes compile errors due to incompatibility with harmcrest 1.4.

**Comment by rocboronat** [comment]
Hello @shuhart! 👋 

Just to clarify: if you used the latest version of Espresso, you would have the same issue, right? And you are using Hamcrest 1.3, aren't you?

Maybe you should move from Hamcrest 1.3 to 2.2 and that's it?

**Comment by rocboronat** [comment]
Sorry! Didn't read you well. You are on 2.2 and Espresso is on 1.3.

Are you on 2.2 for any special reason? May you move to 1.3 or there's any special case 1.3 is not able to handle?

**Comment by shuhart** [comment]
We don't use harmcrest internally. The problem is that Barista depends on harmctest 1.3, that's why the only thing we can do is to downgrade either Espresso or harmcrest (using gradle hacks) or just get rid of Barista.

**Comment by rocboronat** [comment]
Oh, if you don't use Hamcrest, why don't you just remove your Hamcrest dependency? And you could also remove the dependency on Espresso, and just depend on Barista. As Barista depends on both, it's the only dependency you need.

Please, give it a try :·)

**Comment by shuhart** [comment]
Doesn't sound like a good idea. Probably just going to get rid of Barista.

**Comment by rocboronat** [comment]
I think having dependencies you don't use is not too wise, but you choose. Thanks for the feedback, though! 👍

**Comment by Sloy** [comment]
If you're not using Hamcrest, it does sound like a great idea. You should not declare dependencies that you're not going to use. If other libraries require Hamcrest, they will pull the right version already. That's how transitive dependencies work :)

**Comment by shuhart** [comment]
I don't get you. We use Espresso and Barista. And we don't want to
downgrade Espresso just so Barista would work as it depends on the older
version of it.

**Comment by growse** [comment]
@shuhart I think (someone correct me if I'm wrong) the suggestion is to remove the explicit dependency on Espresso, given that Barista already depends on it and will pull it into the tree.

The problem will come if you depend on anything in Espresso in a more recent version than that which is currently pulled in by Barista (v3.1.1 I think). Then it gets a bit awkward.

In my case, I ran into this issue, but then noticed that I'd explicit dependencies on espress which I didn't need. Removing them fixed it.

Is there an existing Barista issue for updating the espresso dependency for those who do need it?

**Comment by rocboronat** [comment]
Now we're working on publishing the latest version of Barista to Android Central. Then, we'll go back to common things like using the latest Espresso version. Sorry for the delay!

By the way, in case @shuhart doesn't do anything special only the latest version of Espresso handles, removing the dependencies and relying on Barista just works 😄

Thanks for your comment, too! Happy you were in the same situation and fixed it just removing your Espresso dependency.

**Comment by shuhart** [comment]
Ok, thanks for the update

**Comment by Sloy** [comment]
Hi @shuhart! Even the latest version of Espresso uses Hamcrest 1.3, so it's not an issue with Barista. Another library must be importing Hamcrest 2.2. 
(According to https://mvnrepository.com/artifact/androidx.test.espresso/espresso-core/3.4.0)

You can check which dependency is pulling Hamcrest 2.2 by running a Gradle scan or using the `dependencies` task as explained here: https://docs.gradle.org/current/userguide/viewing_debugging_dependencies.html

If you remove Barista, you can still have the same error if you use any Espresso matcher that uses a Hamcrest method that doesn't exist in 2.2. I'm not sure that there's a fix for that :/

**Comment by shuhart** [comment]
[espresso-contrib](https://mvnrepository.com/artifact/androidx.test.espresso/espresso-contrib/3.4.0) depends on [accessibility-test-framework](https://mvnrepository.com/artifact/com.google.android.apps.common.testing.accessibility.framework/accessibility-test-framework/3.1) which depends on hamcrest-2.2

**Comment by shuhart** [comment]
We already have a temporary solution for that, I just wanted to point out the issue with you library (was not seeking a solution or advices on how to fix it).

**Comment by Sloy** [comment]
Ah! Good to hear! 
So the fix is to update all Espresso dependencies in Barista, that's easy :)

**Comment by LeMimit** [comment]
Hello,

I got the same problem in my project. it appears that the version of hamcrest is declared strictly to version `2.2` by the last expresso.

Only solution, I found is to downgrade to version `3.3.0` of expresso instead using the version `3.4.0`.

Do you have another solution ? Because this is annoying to use this old version of expresso.

Do you know when it will be fixed on your side ? (Update the version of expresso used by the library)

Thanks !

**Comment by c4augustus** [comment]
This does not fix the essential problem which is actually Google's fault. As pointed out above by @shuhart: 

[espresso-contrib 3.4.0](https://mvnrepository.com/artifact/androidx.test.espresso/espresso-contrib/3.4.0) depends on [accessibility-test-framework 3.1](https://mvnrepository.com/artifact/com.google.android.apps.common.testing.accessibility.framework/accessibility-test-framework/3.1) which depends on hamcrest-2.2

So then the only solution for anything using espresso-contrib, including Barista, is to force the strict downgrade to [accessibility-test-framework 2.1](https://mvnrepository.com/artifact/com.google.android.apps.common.testing.accessibility.framework/accessibility-test-framework/2.1)

This fixes it for me using Espresso 3.4.0 and Barista 4.2.0. Google fix your stuff!

**Comment by RahulSDeshpande** [comment]
So finally after adding this dependency fixed all the errors:
```kt
androidTestImplementation "org.hamcrest:hamcrest:2.2"
```

💃🏼
